db/hints/internal/hint_storage.cc

```diff
@@ -50,16 +50,24 @@ using hints_ep_segments_map = std::unordered_map<unsigned, std::list<fs::path>>;
 // map: IP -> map: shard -> segments
 using hints_segments_map = std::unordered_map<sstring, hints_ep_segments_map>;
 
-future<> scan_for_hints_dirs(const fs::path& hints_directory, std::function<future<> (fs::path dir, directory_entry de, unsigned shard_id)> f) {
-    return lister::scan_dir(hints_directory, lister::dir_entry_types::of<directory_entry_type::directory>(), [f = std::move(f)] (fs::path dir, directory_entry de) mutable {
+/// The caller must ensure that the passed functor's captures
+/// are kept alive at least until the returned future resolves.
+/// It also applies to the functor itself if it's passed as an lvalue reference.
+template <typename Func>
+    requires std::invocable<Func, fs::path, directory_entry, unsigned>
+future<> scan_for_hints_dirs(const fs::path& hint_directory, Func&& func) {
```

Comment:
The function is not on a hot path so I'm not sure it's worth using templates/concepts instead of just std::function.

Reply:
You're correct. I was also wondering if moving to templates might be an overkill here if we end up using I/O anyway. There were two reasons why I decided to go with this solution:

* allowing for functors returning `void`. This could be achieved using `std::function` (or `noncopyable_function`/`std::move_only_function`) too, but we'd end up with a template then anyway,

* though the gain is extremely small, this approach also saves some memory and potential allocations, as well as prevents virtual dispatch. Nowadays, all major compilers use small object optimisation in `std::function` so that it doesn't allocate memory dynamically if the functor's size isn't too big. However, if, for example, the functor has no captures at all, then instead of taking no space (as it could be optimised away by the compiler) or just one byte, we end up with something bigger. The virtual dispatch (being an inherent part of `std::function`) isn't a good thing either as it's more likely that we end up with a cache miss + we need one more level of indirection. Of course, `utils::scan_dir` takes an `std::function` as its argument anyway, so these things will happen, but since we're effectively passing the lambda created in this function's body and the passed functor is only captured by it, we can at least optimise that.

The latter argument isn't very important as I/O will dominate the time this function takes to execute anyway. I mostly had in mind that in the future we might want to reuse the function in other contexts where it could matter more (though probably not).

The first argument might be relevant, but on the other hand, the caller could just wrap their functor in a lambda returning `future<>` and that's it. At the same time, these changes introduce more mental effort on the programmer because they need to ensure that the functor will live at least until the returned future resolves, while with `std::function` we have it, at least in some cases, for free.

I can go back to `std::function`.